library(shiny)
library(plotly)

ui <- fluidPage(
  checkboxInput("use_date_filter", "Use Date Filter", value = FALSE),

  conditionalPanel(
    condition = "input.use_date_filter == true",
    dateRangeInput(
      "dates",
      "Select Date Range:",
      start = NULL,
      end = NULL,
      format = "yyyy-mm-dd"
    ),
  ),
  
  actionButton("refresh", "Refresh Data"),  # Refresh button

  radioButtons("testType", "Workflow", c("SR-WGS" = "WGS", "LR-Amplicon" = "LRS", "WES" = "WES"),inline = TRUE),

  conditionalPanel(
    condition = "input.testType == 'WGS'",
    radioButtons("WGSdataType", "Type", c("Overview" = "wgsoverview", "Coverage" = "wgscoverage", "Mapping" = "wgsmapping", "Variant Calling" = "wgsvariantcalling", "CNVs" = "wgscnv", "SVs" = "wgssv", "Read 1+2" = "wgsr1r2metrics", "Analysis Times" = 'wgstime'),inline = TRUE, selected = "wgsoverview"),
    hr(),
    conditionalPanel(
      condition = "input.WGSdataType == 'wgsoverview'",
      tabsetPanel(
        id = "activeTab_wgsoverview",
        tabPanel("Processed Samples", br(), selected = 0,
                 # two barplots
                 plotlyOutput("barplot_amount_samples_wgs_per_month"), hr(),
                 plotlyOutput("barplot_amount_experiments_wgs_per_month")),
        tabPanel("Duplicate Reads (%)", br(),
                 # one scatterplot with threshold
                 # two boxplots
                 plotlyOutput("scatterplot_threshold_wgs_percentage_duplicates"), hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_percentage_duplicates"), hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_and_stage_percentage_duplicates")),
        tabPanel("Contamination", br(),
                 # one scatterplot with threshold
                 plotlyOutput("scatterplot_threshold_wgs_contamination")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgscoverage'",
      tabsetPanel(
        id = "activeTab_wgscoverage",
        tabPanel("Average Autosomal Coverage", br(),
                 plotlyOutput("scatterplot_threshold_wgs_average_autosomal_coverage"), hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_averageautosomalcoverage"),hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_and_stage_averageautosomalcoverage")), 
        tabPanel("Median Autosomal Coverage", br(),
                 plotlyOutput("scatterplot_threshold_wgs_medianautosomalcoverage")),
        tabPanel("Mean/Median Ratio", br(),
                 plotlyOutput("scatterplot_threshold_wgs_mean_median")),
        tabPanel("Uniformity of coverage (PCT > 0.2*mean)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_uniformity_02")),
        tabPanel("Uniformity of coverage (PCT > 0.4*mean)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_uniformity_04")),
        tabPanel("Mitochondrial", br(),
                 plotlyOutput("scatterplot_threshold_wgs_mitochondrialcoverage")),
        tabPanel("Chromosome x", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_coverage_x")),
        tabPanel("Chromosome y", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_coverage_y")),
        tabPanel("Chromosome", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr1"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr2"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr3"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr4"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr5"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr6"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr7"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr8"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr9"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr10"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr11"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr12"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr13"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr14"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr15"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr16"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr17"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr18"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr19"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr20"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr21"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chr22"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chrx"), hr(),
                 plotlyOutput("scatterplot_nothreshold_wgs_chry")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgsmapping'",
      tabsetPanel(
        id = "activeTab_wgsmapping",
        tabPanel("Input Reads", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_inputreads")),
        tabPanel("Mapped Reads", br(),
                 plotlyOutput("scatterplot_threshold_wgs_percentage_mapped_reads")),
        tabPanel("Q30 Bases", br(),
                 plotlyOutput("scatterplot_threshold_wgs_q30bases"), hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_q30bases"), hr(),
                 plotlyOutput("boxplot_wgs_serialnumber_and_stage_q30bases")),
        tabPanel("Soft-Clipped Bases", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_softclipped")),
        tabPanel("Paired Properly", br(),
                 plotlyOutput("scatterplot_threshold_wgs_properlypaired")),
        tabPanel("Paired dif Chr", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_pairedchromosome")),
        tabPanel("Insert length (Mean)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_insertlength_mean")),
        tabPanel("Insert length (Median)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_insertlength_median")),
        tabPanel("Insert length (std dev)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_insertlength_stddev")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgsvariantcalling'",
      tabsetPanel(
        id = "activeTab_wgsvariantcalling",
        tabPanel("Total Variants", br(),
                 plotlyOutput("scatterplot_threshold_wgs_total_variants")),
        tabPanel("Bi-allelic", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_biallelic")),
        tabPanel("Multi-allelic", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_multiallelic")),
        tabPanel("SNPs (%)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_snps")),
        tabPanel("Ti/Tv ratio", br(),
                 plotlyOutput("scatterplot_threshold_wgs_titv")),
        tabPanel("Het/Hom ratio", br(),
                 plotlyOutput("scatterplot_threshold_wgs_hethom")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgscnv'",
      tabsetPanel(
        id = "activeTab_wgscnv",
        tabPanel("Target Intervals", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_intervals")),
        tabPanel("Segments", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_segments")),
        tabPanel("Amplifications", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_amplifications")),
        tabPanel("Deletions", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_deletions")),
        tabPanel("Passing Amplifications", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_passing_amplifications")),
        tabPanel("Passing Deletions", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_passing_deletions")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgssv'",
      tabsetPanel(
        id = "activeTab_wgssv",
        tabPanel("Total Structural Variants", br(),
                 plotlyOutput("scatterplot_threshold_wgs_structural")),
        tabPanel("SV Deletions (%)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_deletions_sv")),
        tabPanel("SV Insertions (%)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_insertions_sv")),
        tabPanel("SV Duplications (%)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_duplications_sv")),
        tabPanel("SV Breakends (%)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_breakends_sv")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgsr1r2metrics'",
      tabsetPanel(
        id = "activeTab_wgsr1r2metrics",
        tabPanel("Mapped Reads (R1)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_mappedreads_r1")),
        tabPanel("Mapped Reads (R2)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_mappedreads_r2")),
        tabPanel("Q30 Bases (R1)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_q30bases_r1")),
        tabPanel("Q30 Bases (R2)", br(),
                 plotlyOutput("scatterplot_threshold_wgs_q30bases_r2")),
        tabPanel("Soft-Clipped Bases (R1)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_softclippedbases_r1")),
        tabPanel("Soft-Clipped Bases (R2)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_softclippedbases_r2")),
        tabPanel("Mismatched bases (R1)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_mismatchedbases_r1")),
        tabPanel("Mismatched bases (R2)", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_mismatchedbases_r2")),
      ),
    ),
    
    conditionalPanel(
      condition = "input.WGSdataType == 'wgstime'",
      tabsetPanel(
        id = "activeTab_wgstime",
        tabPanel("Total Runtime", br(),
                 plotlyOutput("scatterplot_nothreshold_wgs_runtime")),
      ),
    ),
    hr(),
  ),    
  
  conditionalPanel(
    condition = "input.testType == 'LRS'",
    tabsetPanel(
      id = "activeTab_lrs",
      tabPanel("Total Bases", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_totalbases"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_totalbases"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_totalbases"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_totalbases")),
      tabPanel("P0", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_p0"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_p0"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_p0"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_p0")),
      tabPanel("P1", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_p1"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_p1"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_p1"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_p1")),
      tabPanel("P2", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_p2"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_p2"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_p2"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_p2")),
      tabPanel("Polymerase RL", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_polymerasereadlength"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_polymerasereadlength"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_polymerasereadlength"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_polymerasereadlength")),
      tabPanel("Control Reads", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_controlreads"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_controlreads"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_controlreads"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_controlreads")),
      tabPanel("Control Polymerase RL", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_controlpolymeraserl"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_controlpolymeraserl"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_controlpolymeraserl"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_controlpolymeraserl")),
      tabPanel("Control Concordance", br(),
               plotlyOutput("scatterplot_nothreshold_lrs_controlconcordance"), hr(),
               plotlyOutput("boxplot_lrs_serialnumber_controlconcordance"), hr(),
               plotlyOutput("boxplot_lrs_smrt_cell_lot_number_controlconcordance"), hr(),
               plotlyOutput("boxplot_lrs_sequencing_kit_controlconcordance"))
    ),
    hr(),
  ),

  conditionalPanel(
    condition = "input.testType == 'WES'",
    tabsetPanel(
      id = "activeTab_wes",
      tabPanel("Coverage", br(),
               plotlyOutput("scatterplot_threshold_wes_coverage")),
      tabPanel(">20x Coverage (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_20xcoverage")),
      tabPanel(">50x Coverage (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_50xcoverage")),
      tabPanel("Duplicates (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_duplicates")),
      tabPanel("On Target (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_ontarget"),hr(),
               plotlyOutput("scatterplot_threshold_wes_ontarget_insert")),
      tabPanel("Insert Size", br(),
               plotlyOutput("scatterplot_threshold_wes_insert"),hr(),
               plotlyOutput("scatterplot_threshold_wes_insert_ontarget")),
      tabPanel("Error Rate (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_errorrate")),
      tabPanel("Conifer Segments", br(),
               plotlyOutput("scatterplot_threshold_wes_conifer")),
      tabPanel("GC (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_gc")),
      tabPanel("Contamination (%)", br(),
               plotlyOutput("scatterplot_threshold_wes_contamination")),
      tabPanel("Processed Samples", br(),
               plotlyOutput("barplot_amount_samples_wes_per_month")),
    ),
    hr(),
  ),
)